src/module.c

```diff
@@ -3540,6 +3540,241 @@ int RM_PublishMessageShard(RedisModuleCtx *ctx, RedisModuleString *channel, Redi
+    if (de == NULL) {
+        return;
+    }
+    list *subscribers = dictGetVal(de);
+    listNode *ln = listSearchKey(subscribers, module);
+    if(ln == NULL) {
+        return;
+    }
+    listDelNode(subscribers, ln);
+    if (listLength(subscribers) == 0) {
+        dictDelete(channelModuleSubscribers, channel);
+    }
+    serverLog(LL_DEBUG,"Module %s unsubscribed from channel %s.",
+        module->name, (char *)channel->ptr);
+}
+
+/**
+ * Subscribe the module to a channel with a callback. The API is idempotent based on the channel name, a
+ * module has subscribed to. Hence, if the callback needs to be updated then the client needs to unsubscribe
+ * from the channel and subscribe.
+ * 
+ * The callback registered would be invoked on each message published to a given channel.
+ */
+void RM_SubscribeToChannel(RedisModuleCtx *ctx, RedisModuleString *channel, RedisModuleSubscribeFunc callback) {
+    dictEntry *de = dictFind(moduleChannelSubscribers, ctx->module);
+    RedisModuleChannelSubscriber *moduleChannelSubscriber;
+    if (de == NULL) {
+        moduleChannelSubscriber = zmalloc(sizeof(*moduleChannelSubscriber));
+        moduleChannelSubscriber->subscribe_callback = callback;
+        moduleChannelSubscriber->subscriptions = listCreate();
+        listSetMatchMethod(moduleChannelSubscriber->subscriptions, listMatchStringObjects);
+        listSetFreeMethod(moduleChannelSubscriber->subscriptions, decrRefCountVoid);
+        dictAdd(moduleChannelSubscribers, ctx->module, moduleChannelSubscriber);
+    } else {
+        moduleChannelSubscriber = dictGetVal(de);
+    }
+
+    if (listSearchKey(moduleChannelSubscriber->subscriptions, channel)) {
+        return;
+    }
+
+    /* Add the channel to the list of subscriptions. */
+    robj *channelToAdd = dupStringObject(channel);
+    listAddNodeTail(moduleChannelSubscriber->subscriptions, channelToAdd);
+
+    /* Add the module to the (channel -> list of module subscribers hash table). */
+    list *subscribers = NULL;
+    de = dictFind(channelModuleSubscribers, channelToAdd);
+    if (de == NULL) {
+        subscribers = listCreate();
+        dictAdd(channelModuleSubscribers, channelToAdd, subscribers);
+        incrRefCount(channelToAdd);
+    } else {
+        subscribers = dictGetVal(de);
+    }
+    listAddNodeTail(subscribers, ctx->module);
+
+    serverLog(LL_DEBUG,"Module %s subscribed to channel %s.", ctx->module->name, (char *)channelToAdd->ptr);
+
+    return;
+}
+
+/**
+ * This API unsubscribe(s) the module from the given channel.
+ * 
+ */
+void RM_UnsubscribeFromChannel(RedisModuleCtx *ctx, RedisModuleString *channel) {
+    dictEntry *de = dictFind(moduleChannelSubscribers, ctx->module);
+    if (de == NULL) {
+        return;
+    }
+
+    RedisModuleChannelSubscriber *moduleChannelSubscriber = dictGetVal(de);
+
+    /* No subscription found for module. */
+    if (listLength(moduleChannelSubscriber->subscriptions) == 0) {
```

Comment:
can this be true or this is an error? we remove the module subscriber in case the list is empty right?

Reply:
Changed to assert.